### PR TITLE
[expo-image] Add srcset support to expo-image

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

To improve expo-image on web for static rendering.

# How

We add a new `responsivePolicy`: `static`.
We add a new field to each source called `webMaxViewportWidth`.

If the responsive property is set, the sources with `webMaxViewportWidth` set get mapped to `sizes` and `srcset` properties in .

---

I'd like to keep the sources array as used on native platforms, and instead combine the array into the srcset and sizes tags. This should work for static rendering, and if we want we can change the API to take in the full media query expression instead of just the max viewport width as a number.

I want to keep it explicit that there's a client side responsive images implementation and a server side -> srcset implementation, so for now there are different policies defined in the API. 

We can also make it so that the static policy gets chosen if there's a responsive source with webMaxViewportWidth  so the user does not need to think about it.

# Usage

If you use it like this:

```tsx
<Image
        style={styles.image}
        source={[
          {
            uri: `https://picsum.photos/id/238/800/800`,
            width: 800,
            height: 800,
            webMaxViewportWidth: 800,
          },
          {
            uri: `https://picsum.photos/id/237/1200/1200`,
            width: 1200,
            height: 1200,
            webMaxViewportWidth: 1200,
          },
          {
            uri: `https://picsum.photos/id/236/600/600`,
            width: 600,
            height: 600,
            webMaxViewportWidth: 600,
          },
        ]}
      />
```

It will get mapped to this `` tag:

```html
<img 
src="https://picsum.photos/id/237/1200/1200" 
srcset="https://picsum.photos/id/236/600/600 600w, 
        https://picsum.photos/id/238/800/800 800w, 
        https://picsum.photos/id/237/1200/1200 1200w" 
sizes="(max-width: 600px) 600px, 
       (max-width: 800px) 800px, 
       (max-width: 1200px) 1200px, 
       1200px"
 ...>
```
# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by aleqsio** [comment]
@tsapeta Do you think this `webMaxViewportWidth` property makes sense or should we just take a `mediaQuery` string for each source?

**Comment by pavelbabenko** [comment]
Hi Team!
Could you proceed with this PR? It's quite an essential feature for expo-image

**Review by lukmccall — APPROVED** [review]

File: packages/expo-image/src/Image.types.ts [hunk 1]
```diff
@@ -158,12 +166,19 @@ export interface ImageProps extends ViewProps {
...
+   * Controls the selection of the image source based on the container or viewport size.
+   * If set to `'static'`, the browser selects the correct source based on user's viewport width. Works with static rendering.
+   * Make sure to set the `webMaxViewportWidth` property on all sources.
+   * The source with the largest `webMaxViewportWidth` is used even for larger viewports.
+   *
+   * If set to `'initial'`, the component will select the correct source during mount based on container size. Does not work with static rendering.
+   *
+   * If set to `'live'`, the component will select the correct source on every resize based on container size. Does not work with static rendering.
+   *
+   * @default 'static' if all sources have the `'webMaxViewportWidth'` property set, otherwise 'live'
```
Comment (tsapeta) [inline packages/expo-image/src/Image.types.ts:178]: Some inconsistency here 😅 'static', `static` or `'static'`?

File: packages/expo-image/src/Image.types.ts [hunk 2]
```diff
@@ -42,6 +42,14 @@ export type ImageSource = {
    * If not provided, the `uri` is used also as the cache key.
    */
   cacheKey?: string;
+  /**
+   * @platform web
```
Comment (tsapeta) [inline packages/expo-image/src/Image.types.ts:46]: I think we used to add annotations after the description part

**Review by tsapeta — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]

File: apps/native-component-list/src/screens/Image/ImageScreen.tsx [hunk 3]
```diff
@@ -69,6 +69,13 @@ export const ImageScreens = [
       return optionalRequire(() => require('./ImageBackgroundScreen'));
     },
   },
+  {
+    name: 'Static responsiveness using srcSet',
+    route: 'image/srcset',
+    getComponent() {
+      return optionalRequire(() => require('./ImageSrcSetScreen'));
+    },
+  },
```
[the remark this answers is not in the document]
Comment (aleqsio) [reply apps/native-component-list/src/screens/Image/ImageScreen.tsx:78]: meh, I'd say to keep it in always for our internal demos

File: packages/expo-image/src/ExpoImage.web.tsx [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import React from 'react';
-
+import { View } from 'react-native-web';
```
[the remark this answers is not in the document]
Comment (aleqsio) [reply packages/expo-image/src/ExpoImage.web.tsx]: Actually not, we need both `style` support and the ability to call `getBoundingClientRect` on ref – this is actually the easiest solution now that I think on it.

**Pull request merged**